accord-core/src/main/java/accord/utils/SimpleBitSet.java

```diff
@@ -0,0 +1,348 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package accord.utils;
+
+import java.util.Arrays;
+
+import static java.lang.Long.highestOneBit;
+import static java.lang.Long.lowestOneBit;
+import static java.lang.Long.numberOfTrailingZeros;
+
+public class SimpleBitSet
```

Comment:
I forget if there was another reason (I think there may have been), but I did try and the main one I can remember is that there’s no copy constructor, so you have to explicitly propagate each set bit individually, which would be pretty costly given our CoW semantics.

Reply:
That involves unnecessary array allocations, but we could move to that. Hopefully hotspot will elide the allocations. There's also no equivalent of `prevSetBitNotBefore`, nor any equivalent of supplying the `ifNotFound` which each complicates logic. `forEach` also depends on `IntStream`, so we would be relying on HotSpot to elide these allocations, as well as any capturing lambda (which we explicitly avoid today by accepting `IndexedQuadConsumer`).

Since this is written, is pretty simple and guarantees our requirements I'm strongly inclined to keep it, but I don't mind removing it if you both prefer that.

Reply:
> There's also no equivalent of prevSetBitNotBefore

`java.util.BitSet#nextSetBit` goes forward, `java.util.BitSet#previousSetBit` goes backwards... so you can recreate `prevSetBitNotBefore` without much issue... this would also force `ifNotFound` to match in both cases (if you pass the bound we return `ifNotFound`, and if you pass the block we return `-1` (this looks to be the only function that does this behavior... so feels more like a bug...)...)...

> valueOf also cannot be used by extending implementations, so there's no way to have an ImmutableBitSet, nor one that provides any of these facilities.

Why is that?  You can always extend `BitSet` and block mutable APIs and delegate to read-only apis... Similar in concept to how `java.util.Collections#unmodifiableList` works (proxy reads, block writes)

> Since this is written, is pretty simple and guarantees our requirements I'm strongly inclined to keep it

my mental model of coding is that new code comes with more risk, so if you can do something with existing logic, we should.  We were having a similar conversation in `DeterministicSet` where I pointed out that the development of a new data structure didn't handle all cases correctly which lead to corruption bugs (which I show in the test)... 

So I have a very strong preference to reuse existing rather than reinventing, and have far more concern when we reinvent when we we don't have a strong justification for.

Reply:
> Why is that? You can always extend BitSet and block mutable APIs and delegate to read-only apis... Similar in concept to how java.util.Collections#unmodifiableList works (proxy reads, block writes)

Because `valueOf` cannot be used to construct a subclass of `BitSet`. If the constructor for `BitSet` were public or protected, I would have used it.

> so you can recreate prevSetBitNotBefore

inefficiently

> where I pointed out that the development of a new data structure

I have responded there already. The original data structure did not have this bug, and we were already using this data structure for the purpose. The bug was introduced later, which is a risk in software development. There are action items for expanding test coverage, and likely this would be detected by simulation even if this did not occur. These classes are quite simple, and likely to be thoroughly tested before the project is done.

There is always a question of balancing time investment and outcome, and we are currently pushing for feature completeness, expecting a final pass to clean up minor issues. I therefore do not think it is particularly helpful to point to these kinds of issues as somehow emblematic of anything, when they occur as (or more) often in the *much* more complex code interactions we are layering on top of this, whether we use a standard library or otherwise. This kind of issue is very shallow, and was highly likely to be caught even if we did not write anymore tests. 

I can also point to historic cases where bugs were _included_ in the released software because we _removed_ simple custom data structures by the same argument you make, only to discover the standard library behaviour was surprising and lead to memory leaks. So, this is far from the slam dunk argument people often think it is.

Wasting significant effort debating this kind of ideological point is also highly counterproductive, IMO.

Reply:
> Wasting significant effort debating this kind of ideological point is also highly counterproductive, IMO.

Hey it takes two

I generally prefer stdlib data structures wherever possible, even if the fit is a little awkward. In this case, the lack of a copy constructor is a pretty credible justification, given the frequent switching between mutable and immutable instances.

That said, in retrospect, I think we’ve spent too much time writing new data structures for accord. I understand the motivation, but they’re a real drag on review velocity, since you’re not only painstakingly reasoning about protocol changes, but also painstakingly thinking through large swaths of array math.

Reply:
> if you think there were places I could have made better choices

just want to point out that that wasn't directed at you personally, I've authored some of the structures, and reviewed/approved most of the others

I agree that some level of bespoke logic is unavoidable. But I also think there are a lot of places where we're handling arrays where a library could have provided something sufficient or close enough to avoid mental gymnastics with a negligible/non-existant cpu/memory penalty

Reply:
Potentially, yeah. I think a retrospective would be really valuable in that case, and perhaps a more productive venue for the debate in abstract. I doubt everyone will agree, but we’ll have the benefit of hindsight to discuss what maybe did and didn’t work on each side of the coin.